### PR TITLE
Feature/refactoring cmake structure

### PR DESCRIPTION
**Pull request by BSolonenko**
This PR is **[not ready]** for review.   

### Summary
SDL may use modern cmake approach of creating targets. It will simplify
porting SDL on any platform. For each third party library used by SDL
created .cmake file, that will expose library as cmake_library.

This will allow to make clearer dependencies, avoid dependency gaps
(required for multithreading compilation) and replace dependency during
porting to other platform just with replacing .cmake third party library
file.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -28,13 +28,69 @@
...
+include("${CMAKE_CURRENT_LIST_DIR}/tools/cmake/helpers/install_rules.cmake")
+include("${CMAKE_SOURCE_DIR}/tools/cmake/helpers/platform.cmake")
+include("${CMAKE_CURRENT_LIST_DIR}/tools/cmake/helpers/build_config_generator.cmake")
+
+include(CMakeDependentOption)
+include(GNUInstallDirs)
+
+
+# Request C++11
+#set(CMAKE_CXX_STANDARD 11)
```
Comment (AByzhynar) [inline CMakeLists.txt]: @BSolonenko Commented code?

File: CMakeLists.txt [hunk 2]
```diff
@@ -480,3 +456,5 @@ find_package(Doxygen)
     message(STATUS "To enable processing of MscGen comments please install mscgen")
     message(STATUS "sudo apt-get install mscgen")
 endif()
+
+add_subdirectory(src)
```
Comment (AByzhynar) [inline CMakeLists.txt]: @BSolonenko 
Please add an empty line here

File: cmake/helpers/dependencies.cmake [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+set(dependency_descriptor_dir "${CMAKE_CURRENT_LIST_DIR}/dependencies")
+file(GLOB dependency_descriptors "${dependency_descriptor_dir}/*")
+foreach(dependency IN LISTS dependency_descriptors)
+  get_filename_component(dependency_name "${dependency}" NAME_WE)
+  list(APPEND buildable_dependencies "${dependency_name}")
+endforeach()
+
+function(SDL_find_package)
+  # Handle REQUIRED somehow so this doesn't fail
```
Comment (AByzhynar) [inline cmake/helpers/dependencies.cmake]: @BSolonenko Unclear comment. Please extend...

File: cmake/helpers/dependencies/BSON.cmake [hunk 4]
```diff
@@ -28,19 +28,20 @@
...
-  ${CMAKE_SOURCE_DIR}/src/appMain/test/low_voltage_signals_handler_test.cc
-  ${COMPONENTS_DIR}/utils/test/mock_signals_posix.cc
-)
+include(FetchContent)
 
-set(LIBRARIES
-  gmock
+FetchContent_Declare(
+  BSON
+  GIT_REPOSITORY https://github.com/BSolonenko/bson_c_lib.git
```
Comment (AByzhynar) [inline cmake/helpers/dependencies/BSON.cmake]: @BSolonenko  Don forget to change it before merge

File: cmake/helpers/dependencies/log4cxx.cmake [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+set(log4cxx_INCLUDE_DIR "${THIRD_PARTY_INSTALL_PREFIX}/include/log4cxx/")
+
+find_library(log4cxx_LIBRARY
+  NAMES
+    log4cxx
+  PATH_SUFFIXES
+  lib
+)
```
Comment (AByzhynar) [inline cmake/helpers/dependencies/log4cxx.cmake]: @BSolonenko Align indents..

File: cmake_toolchains/Toolchain-QNX7-QCC-Base.cmake [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+set(CMAKE_SHARED_LIBRARY_SUFFIX ".so")
+set(CMAKE_STATIC_LIBRARY_PREFIX "lib")
+set(CMAKE_STATIC_LIBRARY_SUFFIX ".a")
+
+if(CMAKE_HOST_WIN32)
+    set(HOST_EXECUTABLE_SUFFIX ".exe")
+endif()
+
+if(QNX_HOST)
+    set(ENV{QNX_HOST} "${QNX_HOST}")
```
Comment (AByzhynar) [inline cmake_toolchains/Toolchain-QNX7-QCC-Base.cmake]: @BSolonenko Use same indents over entire code

File: src/3rd_party/CMakeLists.txt [hunk 7]
```diff
@@ -168,80 +170,46 @@ if(ENABLE_LOG)
...
   message(STATUS "Did not find boost. Downloading and installing boost 1.66")
-  set(BOOST_INSTALL_COMMAND ./b2 install)
-  if (${3RD_PARTY_INSTALL_PREFIX} MATCHES "/usr/local")
-    set(BOOST_INSTALL_COMMAND sudo ./b2 install)
+  if(DEFINED CMAKE_TOOLCHAIN_FILE)
+    include("./Boost.cmake")
+  else()
+     set(BOOST_INSTALL_COMMAND ./b2 install)
+     if (${3RD_PARTY_INSTALL_PREFIX} MATCHES "/usr/local")
+       set(BOOST_INSTALL_COMMAND sudo ./b2 install)
```
Comment (AByzhynar) [inline src/3rd_party/CMakeLists.txt]: @BSolonenko Check indents

File: src/3rd_party/apache-log4cxx-cmake/APACHE_LOG4CXX.cmake [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    "OBJCOPY=${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-objcopy${HOST_EXECUTABLE_SUFFIX}"
+    "LINKER=${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-ld"
+    "NM=${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-nm${HOST_EXECUTABLE_SUFFIX}"
+    "RANLIB=${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-ranlib${HOST_EXECUTABLE_SUFFIX}"
+    "AR=${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-ar${HOST_EXECUTABLE_SUFFIX}"
+    "CXX=${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-g++${HOST_EXECUTABLE_SUFFIX}"
+    "CC=${QNX_HOST}/usr/bin/nto${CMAKE_SYSTEM_PROCESSOR}-gcc${HOST_EXECUTABLE_SUFFIX}"
+    )
+
+# set(SIMPLEDATEFORMAT_PATH
```
Comment (AByzhynar) [inline src/3rd_party/apache-log4cxx-cmake/APACHE_LOG4CXX.cmake]: @BSolonenko Uncomment or remove commented code

**Comment by LuxoftAKutsan** [comment]
@BSolonenko  please merge partial implementation, Specify what is currently not working And continue work with separate atomic PRs

**Comment by LuxoftAKutsan** [comment]
@BSolonenko  what is this + 200000 lines about?

**Pull request closed without merging**